sources/ldap: add forward deletion option

gergosimonyi opened the pull request:
Closes #1644 

---

## Checklist

-   [x] Local tests pass (`ak test authentik/`)
-   [x] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [x] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [x] The code has been formatted (`make web`)

If applicable

-   [ ] The documentation has been updated
-   [ ] The documentation has been formatted (`make website`)

gergosimonyi committed:
sources/ldap: add forward deletion option

rissson commented on authentik/sources/ldap/models.py:
```diff
@@ -137,6 +137,15 @@ class LDAPSource(Source):
+        default=False,
+        blank=True,
+        help_text=_(
+            "Deletes authentik users and groups which were previously supplied by this source, "
```
```suggestion
            "Delete authentik users and groups which were previously supplied by this source, "
```

rissson commented on web/src/admin/sources/ldap/LDAPSourceForm.ts:
```diff
@@ -148,6 +148,26 @@ export class LDAPSourceForm extends BaseSourceForm<LDAPSource> {
+                </label>
+                <p class="pf-c-form__helper-text">
+                    ${msg(
+                        "Deletes authentik users and groups which were previously supplied by this source, but are now missing from it.",
```
```suggestion
                        "Delete authentik users and groups which were previously supplied by this source, but are now missing from it.",
```

rissson commented on authentik/sources/ldap/sync/base.py:
```diff
@@ -77,6 +77,12 @@ def get_objects(self, **kwargs) -> Generator:
         """Get objects from LDAP, implemented in subclass"""
         raise NotImplementedError()
 
+    def get_identifier(self, object):
+        attributes = object.get("attributes", {})
+        if not attributes.get(self._source.object_uniqueness_field):
+            return
+        return flatten(attributes[self._source.object_uniqueness_field])
```
This should also be used in other places, like authentik.sources.ldap.sync.users.UserLDAPSynchronizer.sync

gergosimonyi replied:
Yeah, maybe? I did my best, see if you like it.

I'm not sure what that additional `in` check is for (untested code introduced in deb91bd12bdf3fc11807f5d76ea81e883739f358 -- should it be something like an `isinstance(object, dict)` check instead?), but I'm not about to find out, so I did my best to keep the exact same logic for already existing code.

rissson commented on authentik/sources/ldap/tests/test_sync.py:
```diff
@@ -308,3 +314,124 @@ def test_tasks_openldap(self):
         connection = MagicMock(return_value=mock_slapd_connection(LDAP_PASSWORD))
         with patch("authentik.sources.ldap.models.LDAPSource.connection", connection):
             ldap_sync_all.delay().get()
+
```
Can we have a tests that makes sure that if the user exists at the source, it doesn't get deleted?

I suppose the ones above do implicitly test that though

rissson commented on authentik/sources/ldap/models.py:
```diff
@@ -321,6 +330,10 @@ class Meta:
 
 
 class UserLDAPSourceConnection(UserSourceConnection):
+    validated_by = models.UUIDField(
```
should be indexed

rissson commented on authentik/sources/ldap/models.py:
```diff
@@ -335,6 +348,10 @@ class Meta:
 
 
 class GroupLDAPSourceConnection(GroupSourceConnection):
+    validated_by = models.UUIDField(
```
should be indexed

BeryJu reviewed (approved):
Small nits otherwise lgtm

gergosimonyi committed:
Merge branch 'main' into sources/ldap/add-forward-deletion-option

gergosimonyi committed:
remove unnecessary `blank=True`

gergosimonyi committed:
clarify `validated_by` `help_text`

gergosimonyi committed:
add indices to `validated_by`

gergosimonyi committed:
factor out `get_identifier` everywhere and `get_attributes`

I don't know what that additional `in` check is for, but I'm not about
to find out.

gergosimonyi committed:
add tests for known good user and group

gergosimonyi commented on authentik/sources/ldap/models.py:
```diff
@@ -137,6 +137,15 @@ class LDAPSource(Source):
 
+    delete_not_found_objects = models.BooleanField(
+        default=False,
+        blank=True,
```
Thank you, removed.

gergosimonyi commented on authentik/sources/ldap/models.py:
```diff
@@ -321,6 +330,10 @@ class Meta:
 
 class UserLDAPSourceConnection(UserSourceConnection):
+    validated_by = models.UUIDField(
+        null=True, blank=True, help_text=_("Helper field for batch deletions")
```
Done here and in `GroupLDAPSourceConnection`.

gergosimonyi committed:
fixup! add tests for known good user and group

gergosimonyi committed:
fixup! add tests for known good user and group
